firedrake/ensemble/ensemble_function.py

```diff
@@ -0,0 +1,292 @@
+from firedrake.petsc import PETSc
+from firedrake.ensemble.ensemble_functionspace import (
+    EnsembleFunctionSpace, EnsembleDualSpace)
+from firedrake.adjoint_utils import EnsembleFunctionMixin
+from firedrake.function import Function
+from firedrake.norms import norm
+from pyop2 import MixedDat
+
+from functools import cached_property
+from contextlib import contextmanager
+
+__all__ = ("EnsembleFunction", "EnsembleCofunction")
+
+
+class EnsembleFunctionBase(EnsembleFunctionMixin):
+    """
+    A mixed (co)function defined on a :class:`firedrake.Ensemble`.
+    The subcomponents are distributed over the ensemble members, and
+    are specified locally in a :class:`firedrake.EnsembleFunctionSpace`.
+
+    Parameters
+    ----------
+
+    function_space : :class:`firedrake.EnsembleFunctionSpace`.
```

Comment:
I think numpydoc doesn't need rst so `function_space : EnsembleFunctionSpace` will work. Please review the docs to make sure they render properly.